volatility3/framework/plugins/linux/lsblk.py

```diff
@@ -0,0 +1,447 @@
+        return f"{size / 1024 ** 4:.1f}T"
+
+    def _get_type(self, vmlinux, block_device, gendisk):
+        disk_name = utility.array_to_string(gendisk.disk_name)
+
+        if vmlinux.get_type("block_device").has_member("bd_partno"):
+            if block_device.bd_partno:
+                return "part"
+        else:
+            if gendisk.part0.partno:
+                return "part"
+
+        if disk_name.startswith("dm-"):
+            if vmlinux.has_type("mapped_device"):
+                try:
+                    mapped_device = vmlinux.object(
+                        object_type="mapped_device",
+                        offset=gendisk.private_data,
+                        absolute=True,
+                    )
+
+                    hash_cell = vmlinux.object(
+                        object_type="hash_cell",
+                        offset=mapped_device.interface_ptr,
+                        absolute=True,
+                    )
+
+                    if hash_cell.uuid:
+                        return (
+                            utility.pointer_to_string(hash_cell.uuid, 32)
+                            .split("-")[0]
+                            .lower()
+                        )
+                except exceptions.InvalidAddressException:
+                    return "dm"
+            return "dm"
+
+        if disk_name.startswith("loop"):
+            return "loop"
+        elif disk_name.startswith("md"):
+            if vmlinux.has_type("mddev"):
+                mddev = vmlinux.object(
+                    object_type="mddev",
+                    offset=gendisk.private_data,
+                    absolute=True,
+                )
+                try:
+                    return utility.array_to_string(mddev.clevel)
+                except exceptions.InvalidAddressException:
+                    return "md"
+            return "md"
+        elif vmlinux.has_type("scsi_disk") and vmlinux.has_type("scsi_cd"):
+            sr_bdops = vmlinux.object_from_symbol("sr_bdops")
+            sd_fops = vmlinux.object_from_symbol("sd_fops")
+
+            if gendisk.fops == sd_fops.vol.offset:
+                scsi_disk = vmlinux.object(
+                    object_type="scsi_disk",
+                    offset=gendisk.private_data,
+                    absolute=True,
+                )
+                return self._get_scsi_device_type(scsi_disk.device.type)
+
+            elif gendisk.fops == sr_bdops.vol.offset:
+                scsi_cd = vmlinux.object(
+                    object_type="scsi_cd",
+                    offset=gendisk.private_data,
+                    absolute=True,
+                )
+                return self._get_scsi_device_type(scsi_cd.device.type)
+
+        return "disk"
+
+    def _get_scsi_device_type(self, type_code):
+        """
+        SCSI device types.  Copied almost as-is from kernel header
+        (include/scsi/scsi_proto.h)
+        """
+        type_map = {
```

Comment:
This is kinda ok here, but if anything else ever would be able to make use of it, it should be lifted out into `constants.linux` somewhere (if it doesn't already exist as an enumeration in linux symbol tables).